model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/liquibase/MapLiquibaseConnectionProvider.java

```diff
@@ -0,0 +1,111 @@
+import org.jboss.logging.Logger;
+import org.keycloak.connections.jpa.updater.liquibase.conn.LiquibaseConnectionProvider;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.map.storage.jpa.liquibase.extension.GeneratedColumnSqlGenerator;
+import org.keycloak.models.map.storage.jpa.liquibase.extension.CreateJsonIndexChange;
+import org.keycloak.models.map.storage.jpa.liquibase.extension.CreateJsonIndexGenerator;
+import org.keycloak.models.map.storage.jpa.liquibase.extension.GeneratedColumnChange;
+import org.keycloak.models.map.storage.jpa.liquibase.extension.JsonDataType;
+
+/**
+ * A {@link LiquibaseConnectionProvider} implementation for the map-jpa module. This provider registers the custom {@code Liquibase}
+ * changes and data types that were developed to better support working with data stored as JSON in the database.
+ * </p>
+ * An instance of this provider has to be obtained via {@link KeycloakSession#getProvider(Class, String)} as follows:
+ * <pre>
+ *     LiquibaseConnectionProvider liquibaseProvider = session.getProvider(LiquibaseConnectionProvider.class, "map-jpa");
+ * </pre>
+ *
+ * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
+ */
+public class MapLiquibaseConnectionProvider implements LiquibaseConnectionProvider {
+
+    private static final Logger logger = Logger.getLogger(MapLiquibaseConnectionProvider.class);
+
+    private static final AtomicBoolean INITIALIZED = new AtomicBoolean(false);
+
+    public MapLiquibaseConnectionProvider(final KeycloakSession session) {
+        if (! INITIALIZED.get()) {
+            // TODO: all liquibase providers should probably synchronize on the same object.
+            synchronized (INITIALIZED) {
+                if (! INITIALIZED.get()) {
+                    initializeLiquibase();
+                    INITIALIZED.set(true);
+                }
+            }
+        }
+    }
+
+    /**
+     * Registers the custom changes/types so we can work with data stored in JSON format.
+     */
+    protected void initializeLiquibase() {
+
+        // Add custom JSON data type
+        DataTypeFactory.getInstance().register(JsonDataType.class);
+
+        // Add custom change to generate columns from properties in JSON files stored in the DB.
+        ChangeFactory.getInstance().register(GeneratedColumnChange.class);
+        SqlGeneratorFactory.getInstance().register(new GeneratedColumnSqlGenerator());
+
+        // Add custom change to create indexes for properties in JSON files stored in the DB.
+        ChangeFactory.getInstance().register(CreateJsonIndexChange.class);
+        SqlGeneratorFactory.getInstance().register(new CreateJsonIndexGenerator());
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public Liquibase getLiquibase(final Connection connection, final String defaultSchema) {
+        throw new UnsupportedOperationException("Not supported, there's no aggregator changelog in map-jpa");
```

Comment:
This indicates a need for a separate SPI rather than reusing that of model-jpa

Reply:
Yeah, created a new SPI for liquibase connections in the map-jpa module.